### PR TITLE
Email password login

### PR DESCRIPTION
**Pull request by Steven-Jia**

### CODE REVIEW

**Comment by Steven-Jia** [comment]
screen shot 2018-06-10 at 2 46 22 pm
Pixel 2

**Comment by Steven-Jia** [comment]
image
Nexus 5

File: BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/API/BeeTrackerCaller.java
Comment (johnsington) [inline BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/API/BeeTrackerCaller.java:180]: do we have an exception thrown for invalid email addresses as well?

**Comment by Steven-Jia** [comment]
image
login error state

File: BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/API/BeeTrackerCaller.java
Comment (johnsington) [inline BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/API/BeeTrackerCaller.java]: correct me if i'm wrong, but we shouldn't be sending the password's cleartext to the server—at the bare minimum, we should be hashing these values prior to calling this request. this is a big security concern and should be directly addressed

**Comment by Steven-Jia** [comment]
screen shot 2018-06-10 at 2 59 18 pm
Pixel 2



screen shot 2018-06-10 at 3 00 39 pm
Nexus 5

**Review by johnsington — DISMISSED** [review]
I left 2 comments that need to be addressed before merging, please take a look!

**Comment by johnsington** [comment]
Minor visual concern, but why is the text in the Facebook Login button not horizontally centered on the Pixel 2 screenshot?

**Comment by johnsington** [comment]
Is the email input field's active state the red outline/cursor before erroring? If so, this should not be the default case and should be green. If there is an error with the input after the user taps `Log In`, then it turns the outline/cursor red.

File: BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/API/BeeTrackerCaller.java
Comment (Steven-Jia) [reply BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/API/BeeTrackerCaller.java:180]: No but I can make one. Right now, we're not doing anything with these exceptions...

File: BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/API/BeeTrackerCaller.java
Comment (Steven-Jia) [reply BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/API/BeeTrackerCaller.java]: Yeah you're right. I'll fix it and use hashing

Comment (luisasanmartin) [reply BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/API/BeeTrackerCaller.java]: I think the standard way to do this is to use TLS/HTTPS. There's a way to use TLS with retrofit. https://github.com/square/okhttp/wiki/HTTPS We need to sync up with Dinah to see what the backend can handle. I've started the conversation on Facebook.

File: BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/SplashActivity.java
Comment (luisasanmartin) [inline BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/SplashActivity.java]: FYI this was just there for debugging purposes, you can (and probably should) remove it. My bad :)

File: BeeTracker/app/src/main/res/layout/activity_main.xml
Comment (luisasanmartin) [inline BeeTracker/app/src/main/res/layout/activity_main.xml:35]: nit/just curious: why is this sp?

File: BeeTracker/app/src/main/res/values/strings.xml
Comment (luisasanmartin) [inline BeeTracker/app/src/main/res/values/strings.xml]: I was under the impression we were supposed to keep Canadian in the name. I see it's in the designs though so I'll take it up with @johnsington

**Comment by Steven-Jia** [comment]
@johnsington The Facebook Login button with the text was never centered in the button to begin with. I tried for a while and couldn't get it to lay out properly since the centering property applied only to the text and not to the combined text and FB logo. It looks centered on the Nexus 5 because the phone width is smaller, but on larger phones it looks like the Pixel. I'll try wrapping the 2 with a container and center that to see if it works.

Thanks for the note on the input fields' active/error states. I'll fix that.

File: BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/API/BeeTrackerCaller.java
Comment (dinahshi) [reply BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/API/BeeTrackerCaller.java]: Hey! Great comments. I'm happy to share that hashing is more of a server-side concern. In the database we use salts, bcrypt hashing, the whole shebang and it all comes with the Devise gem. For client to server communication, we need to ensure that the password is in the POST body and we use HTTPS. So no hashing required--cleartext is perfect!

**Comment by dinahshi** [comment]
This might save everyone some work: we're not going to be supporting facebook login anymore so we can drop that button completely. (Sorry I didn't make this more clear earlier!) Let's merge this PR first and ask john to rework the spacing in design later.

File: BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/API/BeeTrackerCaller.java
Comment (Steven-Jia) [reply BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/API/BeeTrackerCaller.java:180]: Actually, it seems like the server returns a 401 when the email doesn't exist and this is already being handled in the callback, so I don't think we need an exception for that.

**Comment by Steven-Jia** [comment]
PIxel 2
image

Nexus 5
image

Error with sign-in
image

Active state of underline when error occurs
image

Active state back to green upon typing
image

File: BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/API/BeeTrackerCaller.java
Comment (luisasanmartin) [reply BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/API/BeeTrackerCaller.java:180]: We should probably distinguish between a generic error and a password/email error in the error dialogue to the user. Let's do this as a separate PR. I created an issue for it here: https://github.com/uwblueprint/foe/issues/78

File: BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/MainActivity.java
Comment (luisasanmartin) [inline BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/MainActivity.java]: Can we move this into showErrorDialog? Also please guard the dismiss call by checking if spinningIconDialog is null. We're having an issue with null pointers on the spinning dialog in other places in the app and will be implementing this check throughout.

File: BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/MainActivity.java
Comment (luisasanmartin) [inline BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/MainActivity.java]: I think LoginManager is a Facebook class so it can be removed

File: BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/MainActivity.java
Comment (luisasanmartin) [inline BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/MainActivity.java]: Please check if spinningIconDialog is null (sorry these aren't your changes, my bad again :/ )

File: BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/MainActivity.java
Comment (luisasanmartin) [inline BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/MainActivity.java:134]: Why is this being stored?

File: BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/MainActivity.java
Comment (luisasanmartin) [inline BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/MainActivity.java]: Please remove LoginManager

**Review by luisasanmartin — APPROVED** [review]
Great work Steven :) Thanks for being so timely. I just added a couple of comments, feel free to fix those and merge

File: BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/MainActivity.java
Comment (Steven-Jia) [reply BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/MainActivity.java:134]: The API doc says "This confirms the email. Next we login. This does not necessarily have to be two steps to the user; we can send the API request to /sign_in without asking the user for password again if we have stored it locally".

I decided to store the email/password that the user uses to sign up so we can do the one-step signup -> login, but I'm not sure how the app is supposed to know that a user has confirmed their email. Or maybe Dinah meant the email and password (in secure text mode) would be displayed on the login screen for convenience after they re-open the app after confirming their email, but they still have to press the log in button. Maybe @dinahshi can help clarify?

Comment (dinahshi) [reply BeeTracker/app/src/main/java/com/blueprint/foe/beetracker/MainActivity.java:134]: I was thinking what Steven had in mind: one step signup so the user doesn't have to enter their email or click login. This is where that confirm_success_url would come in; if you can launch the app and trigger the automatic login, it'd be guaranteed to work since the user has just confirmed their email. Or the app could put a login button on the "waiting for confirmation" modal that the user can click once the app is launched (and do nothing if login fails).

We're still figuring out the ux for this so we'll keep you posted. Important thing is if we are storing the password no the device, it should be stored securely.

**Pull request merged**
